keps/sig-apps/2232-jobs-stopped/README.md

```diff
@@ -0,0 +1,544 @@
+	// false to true), the Job controller will delete all Pods associated with
+	// this Job. This is an alpha field and requires enabling the StoppedJob
+	// feature gate. Defaults to false.
+	Stopped *bool `json:"stopped,omitempty"`
+
+	...
+}
+```
+
+As described in the comment, when the boolean is set to true, the
+controller-manager will abstain from creating Pods even if the Job's
+parallelism requirements aren't met. If the Job is already active and is
+updated with `stopped: true`, all Pods owned by this Job will be sent a SIGTERM
+signal and eventually terminated (the Pod's graceful termination period will be
+honoured). Pods terminated this way will be considered a failure and will not
+count towards completions.
+
+Similar to existing [JobConditionType](https://github.com/kubernetes/kubernetes/blob/c98f6bf30890f2c5826067ae50cfc36958106e68/staging/src/k8s.io/api/batch/v1/types.go#L167)s
+"Complete" and "Failed", we propose adding a new condition type called
+"Stopped" that will be a part of the Job's status:
+
+```go
+type JobStatus struct {
+	// The latest available observations of an object's current state. Can be
+	// one of "Stopped", "Failed", or "Complete".
+	Conditions []JobCondition `json:"conditions,omitempty"`
+
+	...
+}
+
+// JobCondition describes current state of a job.
+type JobCondition struct {
+	// Type of job condition, Complete or Failed.
+	Type JobConditionType `json:"type"`
+
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus `json:"status"`
+
+	// Last time the condition transit from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+
+	...
+}
+
+type JobConditionType string
+
+// These are valid conditions of a job.
+const (
+	// JobStopped means the job has been stopped.
+	JobStopped JobConditionType = "Stopped"
+	// JobComplete means the job has completed its execution.
+	JobComplete JobConditionType = "Complete"
+	// JobFailed means the job has failed its execution.
+	JobFailed JobConditionType = "Failed"
+)
+```
+
+To determine if a Job has been stopped, users must look for the JobCondition
+with `Type` as "Stopped" and `Status` as true. If such a JobCondition does not
+exist, the Job has never been stopped. Otherwise, if the `Status` is false, the
+Job was previously stopped and then resumed. Inferring this from the Job spec
+is not recommended as the Job controller may not have seen the update yet. When
+a Job is stopped, the Job controller will set the condition only after all Pods
+are terminating or have been deleted.
+
+The Stopped JobCondition also has a `LastTransitionTime` field. This can be
+used to infer how long a Job has been stopped for (if the `Status` value is
+true).
+
+When a Job is stopped or created in the stopped state, a "Stopped" event will
+be recorded. Similarly, when a Job is resumed from its stopped state, a
+"Resumed" event will be recorded.
```

Comment:
Do we get a Resumed event if the Job is created with stopped=false?

Reply:
No. An event is sent only when the Job transitions between the two states.